### PR TITLE
runners: bossac: Arduino bootloader reset on macOS

### PR DESCRIPTION
**Pull request by galexite**
Fixes compatibility with the Arduino bootloader reset mechanism when
using `west flash` on Darwin hosts.

stty is used to set the serial port in the bossac runner to 1200 baud on
Arduino ATSAMD21 bootloaders. The `-F` argument to stty is a GNU
coreutils addition, and thus it is missing from Darwin's BSD stty, so
change the flag to `-f` when we're not running on Linux.

On Darwin, set DEFAULT_BOSSAC_PORT to None and ask the user to select
one from a list of IOCalloutDevices reported by ioreg, modelled on
get_board_snr from the nrfjprog runner. This is because serial port TTYs
are generated by the device driver, and therefore there is no safe
default.

Tested with an Arduino Nano 33 IoT board.

### CODE REVIEW

File: scripts/west_commands/runners/bossac.py
Comment (mbolivar-nordic) [inline scripts/west_commands/runners/bossac.py:29]: Just a style thing, but can you put the other setting underneath as:

```py
else:
    DEFAULT_BOSSAC_PORT = '/dev/ttyACM0'
```

**Review by mbolivar-nordic — COMMENTED** [review]
Thanks for the fix! I just have a minor style nit, then LGTM.

File: scripts/west_commands/runners/bossac.py
Comment (galexite) [reply scripts/west_commands/runners/bossac.py:26]: Yep! I've amended the commit, I moved this conditional up to where the constant was originally set also.

Comment (mbolivar-nordic) [reply scripts/west_commands/runners/bossac.py:26]: Thanks!

One more question before I approve: can you confirm that '/dev/cu.usbmodem14301' is the same everywhere? I'm confused about the '14301' suffix at the end of the device file name. It looks like a randomly generated or device-specific string and I just want to make sure that all devices actually have the same file name.

For comparison, `/dev/ttyACM0` on Linux is always the first TTY ACM device which was plugged into USB, so it's a safe default.

Comment (galexite) [reply scripts/west_commands/runners/bossac.py:26]: Yes, I'm not so sure about that now. I was trying to find out where I'd read it from, where the number after `usbmodem` is generated somehow from the VID and PID, but I can't find any evidence of that. I think it's probably best if I remove this for now. In theory, the proper way of identifying the port is taking the device node from `ioreg`.

Comment (mbolivar-nordic) [reply scripts/west_commands/runners/bossac.py:26]: Another option is to rework this to prompt the user if there are multiple candidates on each OS. See get_board_snr() in the nrfjprog.py runner file for an example.

**Review by carlescufi — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
> [@carlescufi](/carlescufi) [carlescufi](/carlescufi) requested a review from [mbolivar-nordic](/mbolivar-nordic) [yesterday](#event-4658264202)

My review question is still outstanding. Putting a -1 on so this doesn't get inadvertently merged

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
Awaiting response regarding /dev/cu.usbmodem14301 name

**Comment by galexite** [comment]
Yeah, sorry about that, quite busy at the moment. It would probably be best if that change is omitted now. I don't know, but I think it would be better for a later PR asking the appropriate tty on Linux also regarding the review comment?

**Comment by mbolivar-nordic** [comment]
> Yeah, sorry about that, quite busy at the moment. It would probably be best if that specific change is omitted now.

No worries at all; I just wanted to prevent a merge until it gets sorted.

> I don't know, but I think it would be better for a later PR asking the appropriate tty on Linux also regarding the review comment?

Could be. I think in the Linux case the default is more reasonable, but if there are multiple such devices, adding a prompt makes sense too.

File: scripts/west_commands/runners/bossac.py
Comment (galexite) [reply scripts/west_commands/runners/bossac.py]: OK, should be resolved with this next commit!

**Review by mbolivar-nordic — COMMENTED** [review]
Thanks for the update! Quick question about `ioreg`.

**Review by mbolivar-nordic — APPROVED** [review]
LGTM, but I would like to get @nandojve to approve before this is merged.

**Review by nandojve — APPROVED** [review]
LGTM

**Comment by galexite** [comment]
Should have that fixed for you now!

**Comment by mbolivar-nordic** [comment]
Yep, let's see what CI says; thanks!

**Pull request merged**
